File: app/src/main/java/com/ltst/chessview/sample/MovesAdapter.java
package com.ltst.chessview.sample;

import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.ltst.chessview.ChessView;

import java.util.ArrayList;
import java.util.List;

public class MovesAdapter extends RecyclerView.Adapter<MovesAdapter.ViewHolder> {

    public static final int NO_POSITION = -1;

    private OnSelectedItemChange mOnSelectedItemChange;

    private List<MoveItem> mItems = new ArrayList<>();

    public void setOnSelectedItemChange(@Nullable OnSelectedItemChange onSelectedItemChange) {
        this.mOnSelectedItemChange = onSelectedItemChange;
    }

    public int getSelectedPosition() {
        List<MoveItem> items = getItems();
        for (int i = 0; i < items.size(); i++) {
            if (items.get(i).isSelected()) {
                return i;
            }
        }
        return NO_POSITION;
    }

    public void setSelection(int position) {

        List<MoveItem> items = getItems();

        if (items.size() == 0) return;

        int selectedPosition = getSelectedPosition();

        if (selectedPosition != position) {
            items.get(selectedPosition).setSelected(false);
            notifyItemChanged(selectedPosition);

            items.get(position).setSelected(true);
            notifyItemChanged(position);
        }

        if (mOnSelectedItemChange != null) {
            mOnSelectedItemChange.onSelectedItemChange(position, selectedPosition);
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view = inflater.inflate(R.layout.move_item_view, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder viewHolder, int position) {
        MoveItem item = getItem(position);
        viewHolder.mStep.setVisibility(position % 2 == 0 ? View.VISIBLE : View.GONE);
        viewHolder.mStep.setText(String.valueOf(position / 2 + 1));
        viewHolder.mMove.setText(item.getSan());
        viewHolder.mRoot.setBackground(item.isSelected()
                ? ContextCompat.getDrawable(viewHolder.mRoot.getContext(), R.drawable.sp_rect_accent) : null);
        if (item.getSan().contains("O-O")) {
            viewHolder.mFigure.setVisibility(View.GONE);
        } else {
            ChessView.Cell cell = item.getCell();
            if (cell != null) {
                viewHolder.mFigure.setImageResource(cell.drawable);
            }
            viewHolder.mFigure.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public int getItemCount() {
        return mItems.size();
    }

    public MoveItem getItem(int position) {
        return mItems.get(position);
    }

    @Override
    public long getItemId(int position) {
        return mItems.get(position).getNum();
    }

    private List<MoveItem> getItems() {
        return mItems;
    }

    public void addAll(List<MoveItem> list) {
        mItems.addAll(list);
        notifyDataSetChanged();
    }

    public interface OnSelectedItemChange {
        void onSelectedItemChange(int newPosition, int oldPosition);
    }

    public static final class ViewHolder extends RecyclerView.ViewHolder {

        public final ViewGroup mRoot;
        public final TextView mStep;
        public final ImageView mFigure;
        public final TextView mMove;

        public ViewHolder(View itemView) {
            super(itemView);
            mRoot = (ViewGroup) itemView.findViewById(R.id.board_move_item_root);
            mStep = (TextView) itemView.findViewById(R.id.board_move_item_step);
            mFigure = (ImageView) itemView.findViewById(R.id.board_move_item_figure);
            mMove = (TextView) itemView.findViewById(R.id.board_move_item_cell);
        }
    }
}

File: app/src/main/java/com/cosic/chessview/sample/adapter/MovesAdapter.kt
package com.cosic.chessview.sample.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.cosic.chessview.sample.R
import com.cosic.chessview.sample.bindView

const val NO_POSITION = -1

class MovesAdapter : RecyclerView.Adapter<MovesAdapter.ViewHolder>() {

    var onSelectedItemChange: ((newPosition: Int, oldPosition: Int) -> Unit)? = null

    private val moveItems = mutableListOf<MoveItem>()

    private val items: List<MoveItem>
        get() = moveItems

    val selectedPosition: Int
        get() {
            val items = items
            for (i in items.indices) {
                if (items[i].isSelected) {
                    return i
                }
            }
            return NO_POSITION
        }

    fun setSelection(position: Int) {
        val items = items
        if (items.isEmpty()) return
        val selectedPosition = selectedPosition
        if (selectedPosition != position) {
            items[selectedPosition].isSelected = false
            notifyItemChanged(selectedPosition)
            items[position].isSelected = true
            notifyItemChanged(position)
        }
            onSelectedItemChange?.invoke(position, selectedPosition)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val view = inflater.inflate(R.layout.move_item_view, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(viewHolder: ViewHolder, position: Int) {
        val item = getItem(position)
        viewHolder.mStep.visibility = if (position % 2 == 0) View.VISIBLE else View.GONE
        viewHolder.mStep.text = (position / 2 + 1).toString()
        viewHolder.mMove.text = item.san
        viewHolder.mRoot.background = if (item.isSelected) {
            ContextCompat.getDrawable(viewHolder.mRoot.context, R.drawable.sp_rect_accent)
        } else {
            null
        }
        if (item.san.contains("O-O")) {
            viewHolder.mFigure.visibility = View.GONE
        } else {
            val cell = item.cell
            if (cell != null) {
                viewHolder.mFigure.setImageResource(cell.drawable)
            }
            viewHolder.mFigure.visibility = View.VISIBLE
        }
    }

    override fun getItemCount(): Int {
        return moveItems.size
    }

    fun getItem(position: Int): MoveItem {
        return moveItems[position]
    }

    override fun getItemId(position: Int): Long {
        return moveItems[position].num.toLong()
    }

    fun addAll(list: List<MoveItem>) {
        moveItems.addAll(list)
        notifyDataSetChanged()
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val mRoot by bindView<ViewGroup>(R.id.board_move_item_root)
        val mStep by bindView<TextView>(R.id.board_move_item_step)
        val mFigure by bindView<ImageView>(R.id.board_move_item_figure)
        val mMove by bindView<TextView>(R.id.board_move_item_cell)
    }
}